examples/structured_data/movielens_recommendations_transformers.py

```diff
@@ -0,0 +1,608 @@
+"""
+Title: A recommendation system with a transformer
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2020/12/30
+Last modified: 2020/12/30
+Description: Using the Behavior Sequence Transformer model for on Movielens dataset.
+"""
+
+"""
+## Introduction
+
+This example demonstrates the [Behavior Sequence Transformer (BST)](https://arxiv.org/abs/1905.06874)
+model, by Qiwei Chen et al., using the [Movielens dataset](https://grouplens.org/datasets/movielens/).
+The BST model leverages the sequential behaviour of the users in watching and rating movies,
+as well as user profile and movie features, to predict the rating of the user to a target movie.
+
+More precisely, the BST model aims to predict the rating of a target movie by accepting
+the following inputs:
+1. A fixed-length *sequence* of `movie_ids` watched by a user.
+2. A fixed-length *sequence* of the `ratings` for the movies watched by a user.
+3. A set of user features, including `user_id`, `sex`, `occupation`, and `age_group`.
+4. A set of `genres` for each movie in the input sequence and the target movie.
+5. A `target_movie_id` for which to predict the rating.
+
+This example modifies the original BST model in the following ways:
+1. We incorporate the movie features (genres) into the processing of the embedding of each
+movie of the input sequence and the target movie, rather than treating them as "other features" outside the transformer layer.
+2. We utilize the ratings of movies in the input sequence, along with the their positions
+in the sequence, to update them before feeding them into the self-attention layer.
+
+
+Note that this example should be run with TensorFlow 2.4 or higher.
+"""
+
+"""
+## The dataset
+
+We use the [1M version of the Movielens dataset](https://grouplens.org/datasets/movielens/1m/).
+The dataset includes around 1 million ratings from 6000 users on 4000 movies, along with some user features, movie genres. In addition, the timestamp of each user-movie rating is provided, which allows creating sequences of movie ratings for each user, as expected by the BST model.
+"""
+
+"""
+## Setup
+"""
+
+import os
+import math
+from zipfile import ZipFile
+from urllib.request import urlretrieve
+import pandas as pd
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+from tensorflow.keras.layers.experimental.preprocessing import StringLookup
+from sklearn.model_selection import train_test_split
```

Comment:
Do this by hand, don't import sklearn just for this

Reply:
Done